### PR TITLE
cmake: Remove legacy MSVC build system

### PR DESCRIPTION
**Pull request by hebasto**
At this stage, the new CMake-based build system offers a comprehensive set of features, surpassing those of the legacy build system located in the `build_msvc` directory.

As part of the transition to the new build system, the legacy MSVC build system has been removed.

Additionally, a new ["Windows / MSVC Build Guide"](https://github.com/hebasto/bitcoin/blob/240309-cmake-BX/doc/build-windows-msvc.md) has been introduced, based on https://gist.github.com/hebasto/32c77c0cd592d13685437271ac6b79f6. Notably, the guide lacks information regarding the disabling of base address randomization (see https://github.com/bitcoin/bitcoin/pull/21045). Even with `-DHARDENING=OFF`, this feature remains enabled by default.

### CODE REVIEW

**Comment by hebasto** [comment]
Friendly ping @sipsorcery @pablomartin4btc @m3dwards @TheCharlatan @vasild :)

Also cc @EthanHeilman regarding the base address randomization feature.

**Comment by EthanHeilman** [comment]
@hebasto I'm not sure what exactly what the ask is on the base address randomization feature, but gcc has its own ways of doing it. If msvc is removed, that feature has no purpose.

I'm probably the only person that regularly builds bitcoin on msvc. I'll be a little a little sad to see it go, but it will let us clean up some macros in the RNG.

**Comment by EthanHeilman** [comment]
--disable-hardening doesn't work anymore? Is this just a windows think or is it broken on linux as well?

I ran into [issues with this on OSX](https://twitter.com/Ethan_Heilman/status/1355930152342085633):

```
To disable PIE in bitcoind on OSX run:
make clean
make all LDFLAGS="$LDFLAGS -Wl,-no_pie" 

to check it worked (compare imgs below):
otool -hv src/bitcoind  
```

**Comment by hebasto** [comment]
> --disable-hardening doesn't work anymore? Is this just a windows think or is it broken on linux as well?
> 
> I ran into [issues with this on OSX](https://twitter.com/Ethan_Heilman/status/1355930152342085633):
> 
> ```
> To disable PIE in bitcoind on OSX run:
> make clean
> make all LDFLAGS="$LDFLAGS -Wl,-no_pie" 
> 
> to check it worked (compare imgs below):
> otool -hv src/bitcoind  
> ```

On the master branch, while `--disable-hardening` means "do not attempt to harden the resulting executables", which implies it does _not_ force non-hardening preprocessor macros, compiler and linker flags.

The new CMake-based build system preserves that behaviour.

**Comment by EthanHeilman** [comment]
> On the master branch, while --disable-hardening means "do not attempt to harden the resulting executables", which implies it does not force non-hardening preprocessor macros, compiler and linker flags.
> The new CMake-based build system preserves that behaviour.

So if we wanted the ability to disable windows base address randomization we would need to add it as a CMake flag?

**Comment by hebasto** [comment]
> > On the master branch, while --disable-hardening means "do not attempt to harden the resulting executables", which implies it does not force non-hardening preprocessor macros, compiler and linker flags.
> > The new CMake-based build system preserves that behaviour.
> 
> So if we wanted the ability to disable windows base address randomization we would need to add it as a CMake flag?

Yes. Something like that:
```
cmake -B build --preset vs2022 -DHARDENING=OFF -DCMAKE_EXE_LINKER_FLAGS="/DYNAMICBASE:NO"
```

I'm still curios about the use cases for binaries (1) with disabled ASLR and (2) with disabled PIE. Is it worth considering adding dedicated build configuration options for such cases?

**Comment by EthanHeilman** [comment]
> I'm still curios about the use cases for binaries (1) with disabled ASLR and (2) with disabled PIE.

My main usecase is that bitcoin uses process memory as a source for randomness for it's RNG. In measuring and testing the randomnesss provided by the RNG, I turn all this off to make it is easier to isolate randomness problems. This can also be useful when attempting to replicate heisenbugs. That's a very boutique use case though,

> Is it worth considering adding dedicated build configuration options for such cases?

I'll probably add them back in in a PR when restarting the RNG measurement project. It would save me some time if someone else readded them, but given that as far as I am aware I am the primary user of this feature it seems a little unfair to ask someone else to do the work.

**Comment by hebasto** [comment]
Rebased.

**Comment by hebasto** [comment]
Rebased.

File: doc/build-windows-msvc.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ctest --build-config Release
+```
+
+### 5. Building with Static Linking
+
+```
+cmake -S .. --preset vs2022-static  # It might take a while if the vcpkg binary cache is unpopulated or invalidated
+cmake --build . --config Release
+ctest --build-config Release
+cmake --install . --config Release  # Optional
```
Comment (m3dwards) [inline doc/build-windows-msvc.md:70]: Is there a reason why `install` is listed as part of the static linking build and not the dynamic linking build?

File: doc/build-windows-msvc.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+```
+
+## Performance Notes
+
+### 6. vcpkg Manifest Default Features
+
+One can skip vcpkg manifest default features to speedup the configuration step.
+For example, the following invocation will skip all features except for "wallet" and "tests" and their dependencies:
+```
+cmake -S .. --preset vs2022 -DVCPKG_MANIFEST_NO_DEFAULT_FEATURES=ON -DVCPKG_MANIFEST_FEATURES="wallet;tests"
```
[the remark this answers is not in the document]
Comment (hebasto) [reply doc/build-windows-msvc.md:80]: The features are listed in the `<root>\vcpkg.json` file. What wording would you suggest to list them in the docs?

**Comment by m3dwards** [comment]
Should the guide mention ccache?

What about only building release versions of dependencies with:

```shell
Add-Content -Path "$env:VCPKG_ROOT\triplets\x64-windows.cmake" -Value "set(VCPKG_BUILD_TYPE release)"
Add-Content -Path "$env:VCPKG_ROOT\triplets\${{ matrix.conf.triplet }}.cmake" -Value "set(VCPKG_BUILD_TYPE release)"
```

File: doc/build-windows-msvc.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (m3dwards) [reply doc/build-windows-msvc.md:80]: Available features are listed in the `<root>\vcpkg.json` file.

**Comment by m3dwards** [comment]
I've found that when I change configuration I have to delete the build dir and start again with a fresh one. I assume this is normal and if it is, should this be mentioned?

**Comment by m3dwards** [comment]
If I run:

```shell
mkdir build
cd build
cmake -S .. --preset vs2022-static
cmake -S .. --preset vs2022 -DVCPKG_MANIFEST_NO_DEFAULT_FEATURES=ON -DVCPKG_MANIFEST_FEATURES="wallet;tests"
```

I get:

```shell
-- Configuring incomplete, errors occurred!
```

So moving between static and dynamic presets for example.

**Comment by m3dwards** [comment]
ACK https://github.com/hebasto/bitcoin/commit/8d3bb2beac4ecf4fc6f8a79c78b22a873e86561a

Tested with PowerShell on Windows 11 but I want to give the caveat that these steps were not run on a clean install of windows. I do intend to wipe and reinstall my windows box soon and so will go through the steps again.

**Comment by hebasto** [comment]
> If I run:
> 
> ```shell
> mkdir build
> cd build
> cmake -S .. --preset vs2022-static
> cmake -S .. --preset vs2022 -DVCPKG_MANIFEST_NO_DEFAULT_FEATURES=ON -DVCPKG_MANIFEST_FEATURES="wallet;tests"
> ```
> 
> I get:
> 
> ```shell
> -- Configuring incomplete, errors occurred!
> ```
> 
> So moving between static and dynamic presets for example.
> 
> If we think that's it's perfectly obvious that this would be expected then no probs. I did expect it as it's common with autotools / make when running the configure step again without a distclean.

Right. When using CMake 3.24 and newer, which is usually the case on Windows, the [`--fresh`](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-fresh) command-line option is available as an alternative.

**Pull request merged**